src/data/polynomial/lifts.lean

```diff
@@ -0,0 +1,315 @@
+  by use C r; rw [map_C]
+
+/--For any `(n : ℕ)`, the polynomial `(n : polynomial S)` lifts. -/
+lemma lifts.nat_mem (f : R →+* S) (n : ℕ) : lifts f (n : polynomial S) :=
+  by rw [lifts_iff]; exact subsemiring.coe_nat_mem (ring_hom.of (map f)).srange n
+
+/--The polynomial `X` lifts. -/
+lemma X_lifts (f : R →+* S) : lifts f (X : polynomial S) :=
+  by use X; rw [map_X]
+
+/--The polynomial `X ^ n` lifts. -/
+lemma X_pow_lifts (f : R →+* S) (n : ℕ) : lifts f (X ^ n : polynomial S) :=
+  by use X ^ n; rw [map_pow, map_X]
+
+/--If `p` and `q` lift then `p + q` lifts. -/
+lemma lifts_add {p q : polynomial S} (hp : lifts f p) (hq : lifts f q) : lifts f (p + q) :=
+  by rw lifts_iff at hp hq ⊢; exact subsemiring.add_mem (ring_hom.of (map f)).srange hp hq
+
+/--If `p` and `q` lift then `p * q` lifts. -/
+lemma lifts_mul {p q : polynomial S} (hp : lifts f p) (hq : lifts f q) : lifts f (p * q) :=
+  by rw lifts_iff at hp hq ⊢; exact subsemiring.mul_mem (ring_hom.of (map f)).srange hp hq
+
+/--If `p` lifts and `(n : ℕ)` then `p ^ n` lifts. -/
+lemma lifts.pow {p : polynomial S} (n : ℕ) (hp : lifts f p) : lifts f (p ^ n) :=
+  by rw lifts_iff at hp ⊢; exact subsemiring.pow_mem (ring_hom.of (map f)).srange hp n
+
+/--If `p` lifts and `(n : ℕ)` then `n • p` lifts. -/
+lemma lifts.nat_smul {p : polynomial S} (n : ℕ) (hp : lifts f p) : lifts f (n • p) :=
+  by rw lifts_iff at hp ⊢; exact subsemiring.nsmul_mem (ring_hom.of (map f)).srange hp n
+
+/--If `p` lifts and `(r : R)` then `r * p` lifts. -/
+lemma lifts.base_mul {p : polynomial S} (r : R) (hp : lifts f p) : lifts f (C (f r) * p) :=
+  by rw lifts at hp ⊢; obtain ⟨p₁, rfl⟩ := hp; use C r * p₁; rw [map_mul, map_C]
+
+/--If any element of a list of polynomials lifts, then the product lifts-/
+lemma lifts_list_prod {L : list (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ L → lifts f p) → lifts f L.prod :=
+  by simp_rw [lifts_iff]; exact subsemiring.list_prod_mem (ring_hom.of (map f)).srange
+
+/--If any element of a list of polynomials lifts, then the sum lifts-/
+lemma lifts_list_sum {L : list (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ L → lifts f p) → lifts f L.sum :=
+  by simp_rw [lifts_iff]; exact subsemiring.list_sum_mem (ring_hom.of (map f)).srange
+
+/--If any element of a multiset of polynomials lifts, then the sum lifts-/
+lemma lifts_multiset_sum {m : multiset (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ m → lifts f p) → lifts f m.sum :=
+  by simp_rw [lifts_iff]; exact subsemiring.multiset_sum_mem (ring_hom.of (map f)).srange m
+
+/--If any element of a finset of polynomials lifts, then the sum lifts-/
+lemma lifts_sum {ι : Type w} {t : finset ι} {g : ι → (polynomial S)} :
+  (∀ (x : ι), x ∈ t → lifts f (g x)) → lifts f (∑ (x : ι) in t, g x) :=
+by simp_rw [lifts_iff]; exact subsemiring.sum_mem (ring_hom.of (map f)).srange
+
+/--If `p` lifts then any `monomial n (p.coeff n)` lifts. -/
+lemma monom_lifts_of_lifts {p : polynomial S} (n : ℕ) (h : lifts f p) :
```

Comment:
This lemma seems a bit out of place to me. I would expect `lifts f (monomial n c)` and as argument `h : \ex a, f a = c`, or something like that. Maybe `c \in set.range f` or some variant.

Reply:
I've hadded  `lifts_monomial` and `lifts_iff_coeff_lifts` and removed  `monom_lifts_of_lifts`.